pkg/cloudprovider/aws.go

```diff
@@ -1,32 +1,342 @@
+				}
+			}
+			if len(deleteIPs) == 0 {
+				return NonExistingIPError
+			}
+			input := ec2.UnassignIpv6AddressesInput{
+				NetworkInterfaceId: networkInterface.NetworkInterfaceId,
+				Ipv6Addresses:      deleteIPs,
+			}
+			_, err = a.client.UnassignIpv6Addresses(&input)
+			if err != nil {
+				return err
+			}
+			return a.waitForCompletion(node, awsapi.StringValueSlice(deleteIPs), true)
+		}
+		for _, assignedIPv4 := range networkInterface.PrivateIpAddresses {
+			if assignedIP := net.ParseIP(*assignedIPv4.PrivateIpAddress); assignedIP != nil && assignedIP.Equal(ip) {
+				deleteIPs = append(deleteIPs, assignedIPv4.PrivateIpAddress)
+			}
+		}
+		if len(deleteIPs) == 0 {
+			return NonExistingIPError
+		}
+		inputV4 := ec2.UnassignPrivateIpAddressesInput{
+			NetworkInterfaceId: networkInterface.NetworkInterfaceId,
+			PrivateIpAddresses: deleteIPs,
+		}
+		_, err = a.client.UnassignPrivateIpAddresses(&inputV4)
+		if err != nil {
+			return err
+		}
+		return a.waitForCompletion(node, awsapi.StringValueSlice(deleteIPs), true)
+	}
 	return nil
 }
 
 func (a *AWS) GetNodeEgressIPConfiguration(node *corev1.Node) ([]*NodeEgressIPConfiguration, error) {
+	instance, err := a.getInstance(node)
+	if err != nil {
+		return nil, err
+	}
+	instanceV4Capacity, instanceV6Capacity, err := a.getInstanceCapacity(instance)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving the instance capacities, err: %v", err)
+	}
+	networkInterfaces, err := a.getNetworkInterfaces(instance)
+	if err != nil {
+		return nil, err
+	}
+	for _, networkInterface := range networkInterfaces {
+		config := &NodeEgressIPConfiguration{
+			Interface: *networkInterface.NetworkInterfaceId,
+		}
+		v4Subnet, v6Subnet, err := a.getSubnet(networkInterface)
+		if err != nil {
+			return nil, fmt.Errorf("error retrieving the network interface subnets, err: %v", err)
+		}
+		config.IFAddr = ifAddr{}
+		if v4Subnet != nil {
+			config.IFAddr.IPv4 = v4Subnet.String()
+		}
+		if v6Subnet != nil {
+			config.IFAddr.IPv6 = v6Subnet.String()
+		}
+		capV4, capV6 := a.getCapacity(instanceV4Capacity, instanceV6Capacity, networkInterface)
+		config.Capacity = capacity{
+			IPv4: capV4,
+			IPv6: capV6,
+		}
+		return []*NodeEgressIPConfiguration{config}, nil
+	}
 	return nil, nil
 }
+
+// Unfortunately the AWS API (WaitUntilInstanceRunning) only handles equality
+// assertion: so on delete we can't specify and assert that the IP which is
+// being removed is completely removed, we are forced to do the inverse, i.e:
+// assert that all IPs except the IP being removed are there (which should
+// anyways be the case). Hence, use our own poller which verifies on ADD that
+// all IPs that we add are assigned to the node, and on DEL that all IPs being
+// removed have been completely removed from the node.
+func (a *AWS) waitForCompletion(node *corev1.Node, ips []string, deleteOp bool) error {
+	return wait.PollImmediate(time.Second*2, time.Minute, func() (done bool, err error) {
```

Comment:
it's generally preferred to use `wait.ExponentialBackoff`

Reply:
I thought it would be more straightforward to have a fixed frequency for polling with a specified deadline. `ExponentialBackoff` would make things a lot less non-obvious w.r.t how we are polling.